test/Registry.test.ts

```diff
@@ -0,0 +1,193 @@
+import { ethers } from "hardhat";
+import chai from "chai";
+const { expect } = chai;
+import { generateHexString } from "@unfinishedlabs/test-generators"
+
+describe("Registry", () => {
+  const handle = "flarp";
+  let signer1, signer2, signer3;
+  let delegate1, delegate2, delegate3, newDelegate1;
+
+  let registry;
+
+  beforeEach(async () => {
+    const signers = await ethers.getSigners();
+    signer1 = signers[0];
+    signer2 = signers[1];
+    signer3 = signers[2];
+
+    const Registry = await ethers.getContractFactory("Registry");
+    registry = await Registry.deploy();
+    await registry.deployed();
+
+    const TestDelegate = await ethers.getContractFactory("TestDelegate");
+    delegate1 = await TestDelegate.deploy(signer1.address);
+    await delegate1.deployed();
+
+    newDelegate1 = await TestDelegate.deploy(signer1.address);
+    await newDelegate1.deployed();
+
+    delegate2 = await TestDelegate.deploy(signer2.address);
+    await delegate2.deployed();
+
+    delegate3 = await TestDelegate.deploy(signer3.address);
+    await delegate3.deployed();
+  });
+
+  describe("register", () => {
+
+    it("emits a DSNPRegistryUpdate event", async () => {
+      await expect(registry.connect(signer1).register(delegate1.address, handle))
+        .to.emit(registry, "DSNPRegistryUpdate")
+        .withArgs(1, delegate1.address, handle);
+    });
+
+    it("reverts when sender is not authorized", async () => {
+      await expect(registry.connect(signer2).register(delegate1.address, handle))
+        .to.be.revertedWith("Access denied");
+    })
+
+    it("reverts when contract does not exist", async () => {
+      const bogusContract = ethers.utils.getAddress(generateHexString(40))
+      await expect(registry.connect(signer2).register(bogusContract, handle))
+        .to.be.revertedWith("function call to a non-contract account");
```

Comment:
I can think four potential error modes:
1. The specified address doesn't exist. This fails with a reasonable message.
2. The specified address points to a contract that can't handle the signature or returns the wrong time. This also fails with reasonable message.
3. The specified address handles the signature and returns `true`. This succeeds. If it consistently returns true it can be fixed. Otherwise the handle is locked out.
4. The specified address is an IDelegate, but its the wrong contract or a misconfigured one. This fails because it will return false. There is a possibility that the address exists in the delegate but the delegate itself is not under the sender's control. This could result in the sender losing the handle.

So there is a user-error case that, in very unlucky circumstances, results in a handle being locked out. This could be fixed with a magic number. I suspect that the handles that get locked out in this way will be massively dwarfed by the handles that get locked by users losing access to their delegate, so I'm not sure if it's worth the effort.